### PR TITLE
Fixes #27: Add Customizable Input Component.

### LINKED ISSUE #27: [Feature Request]: Custom input with label
### Is your feature request related to a problem? Please describe.

We need custom input with label and without label too.

There are various types of input across the website so do navigate the website properly.
For example:
https://homy-next.vercel.app/dashboard/profile
https://homy-next.vercel.app/ (click on login)
https://homy-next.vercel.app/listing_details_01 (scroll down to see inputs on right side).

Screenshot 2024-08-22 at 12 10 22 PM
Screenshot 2024-08-22 at 12 11 05 PM
Screenshot 2024-08-22 at 12 12 39 PM

### Describe the solution you'd like

For now we want to accommodate above 3 inputs only. If you find any other type of input, mention it in PR or on this issue so that we can discuss if we are ignoring it or not.

The name of the component should be simply `Input`

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by aniketpatidar**
Fixes #27 

What type of PR is this?
- [x] Feature

### Description
This PR introduces a customizable input component for form inputs. The component allows you to easily modify styles using props.

Example usage

```
<Input
  label="Your Name"
  name="fullName"
  placeholder="Your full name"
  required
  labelColor="text-red-500"
/>

<Input
  label="Email"
  name="email"
  type="email"
  placeholder="youremail@gmail.com"
  required
  labelFontWeight="font-bold"
  labelClassName="text-green-500"
/>

<Input
  label="Address"
  name="address"
  placeholder="Your address"
  borderColor="border-black-300"
  focusBorderColor="focus:border-blue-500"
  backgroundColor="bg-blue-50"
/>
```

### CODE REVIEW

**Comment by aniketpatidar** [comment]
So each of the three strings has its own set of classes? which are fixed for each and cannot be changed from where we are using the input component from where we only pass the type of input component (default small or secondary)

**Comment by rt4914** [comment]
> So each of the three strings has its own set of classes? which are fixed for each and cannot be changed from where we are using the input component from where we only pass the type of input component (default small or secondary)

Yes based on these strings you can control the design/colors in Input.jsx file.
But if no string is passed (DEFAULT) string will get used, which has its own properties.

**Comment by aniketpatidar** [comment]
> > So each of the three strings has its own set of classes? which are fixed for each and cannot be changed from where we are using the input component from where we only pass the type of input component (default small or secondary)
> 
> Yes based on these strings you can control the design/colors in Input.jsx file. But if no string is passed (DEFAULT) string will get used, which has its own properties.
Okay I will do this changes.

**Comment by aniketpatidar** [comment]
```
  <Input
    label="Your Name"
    name="name"
    type="text"
    placeholder="Your full name"
    required
  />
  <Input
    label="Email"
    name="email"
    type="email"
    placeholder="Youremail@gmail.com"
    required
    variant="SECONDARY"
  />
  <Input
    label="Username"
    name="username"
    type="text"
    placeholder="JonyRio"
    required
    variant="SMALL"
  />
```
   Screenshot from 2024-09-10 01-32-09

**Review by rt4914 — CHANGES_REQUESTED** [review]
Suggested changes.

File: src/components/Input.jsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
+const Input = ({
+  label,
+  name,
+  type = 'text',
+  placeholder,
+  required = false,
+  variant = 'DEFAULT'
```
[the remark this answers is not in the document]
Comment (aniketpatidar) [reply src/components/Input.jsx:7]: Thank you for the suggestion! will implement an enum for the variant prop.

File: src/components/Input.jsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    SMALL: {
+      input: 'text-sm border-gray-300 text-gray-700',
+      label: 'font-semibold text-xs',
+    },
+  };
+
+  const { input: inputStyles, label: labelStyles } = variants[variant] || variants.DEFAULT;
+
+  return (
+    <div className="mb-4">
```
[the remark this answers is not in the document]
Comment (aniketpatidar) [reply src/components/Input.jsx:27]: Okay, I will remove that line from the component.

File: src/components/Input.jsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
+const Variant = {
+  DEFAULT: 'DEFAULT',
+  SECONDARY: 'SECONDARY',
+  SMALL: 'SMALL',
+};
+
+const Input = ({
+  label,
```
Comment (rt4914) [inline src/components/Input.jsx:8]: Lets add prop validations too to make sure that `label`, `name` and `placeholder` are always required and also define the proptypes for other attributes.

File: src/components/Input.jsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
+const Variant = {
```
Comment (rt4914) [inline src/components/Input.jsx:1]: Shift this to the bottom of this file just above export.

**Review by rt4914 — CHANGES_REQUESTED** [review]
Really close to getting merged. Great work.

File: src/components/Input.jsx (hunk 3 above)
Comment (aniketpatidar) [reply src/components/Input.jsx:8]: Will do the changes.

**Review by rt4914 — APPROVED** [review]
@aniketpatidar Really great work.
If there is any future component related PR done by someone else, if you want you can add your review on that PR.
It will still need my review at the end.

**Comment by aniketpatidar** [comment]
> aniketpatidar Really great work.
> If there is any future component related PR done by someone else, if you want you can add your review on that PR.
> It will still need my review at the end.

Thank you! Sure, will do!

**Pull request merged**